engine/flink/avro-util/src/main/scala/pl/touk/nussknacker/engine/avro/source/SpecificRecordKafkaAvroSourceFactory.scala

```diff
@@ -1,46 +1,74 @@
+import pl.touk.nussknacker.engine.avro.schemaregistry.{LatestSchemaVersion, SchemaRegistryProvider, SchemaVersionOption, SpecificRecordEmbeddedSchemaDeterminer}
+import pl.touk.nussknacker.engine.avro.source.KafkaAvroSourceFactory.KafkaAvroSourceFactoryState
+import pl.touk.nussknacker.engine.flink.api.process.FlinkSource
 import pl.touk.nussknacker.engine.flink.api.timestampwatermark.TimestampWatermarkHandler
-import pl.touk.nussknacker.engine.kafka.source.KafkaSource
-import pl.touk.nussknacker.engine.kafka.{KafkaConfig, KafkaUtils}
+import pl.touk.nussknacker.engine.kafka.PreparedKafkaTopic
 
 import scala.reflect._
 
 /**
  * Source factory for specific records - mainly generated from schema.
  */
-class SpecificRecordKafkaAvroSourceFactory[T <: SpecificRecord: ClassTag](schemaRegistryProvider: SchemaRegistryProvider,
-                                                                          processObjectDependencies: ProcessObjectDependencies, timestampAssigner: Option[TimestampWatermarkHandler[T]])
-  extends BaseKafkaAvroSourceFactory[T](timestampAssigner) {
+class SpecificRecordKafkaAvroSourceFactory[V <: SpecificRecord: ClassTag](schemaRegistryProvider: SchemaRegistryProvider,
+                                                                          processObjectDependencies: ProcessObjectDependencies,
+                                                                          timestampAssigner: Option[TimestampWatermarkHandler[ConsumerRecord[Any, V]]])
+  extends KafkaAvroSourceFactory[Any, V](schemaRegistryProvider, processObjectDependencies, timestampAssigner) {
 
-  // TODO: it should return suggestions for topics like it is in generic version (KafkaAvroSourceFactory)
-  @MethodToInvoke
-  def createSource(@ParamName(`TopicParamName`) @NotBlank @SimpleEditor(`type` = SimpleEditorType.STRING_EDITOR) topic: String)
-                  (implicit processMetaData: MetaData, nodeId: NodeId): KafkaSource[T] = {
-    val kafkaConfig = KafkaConfig.parseProcessObjectDependencies(processObjectDependencies)
-    val preparedTopic = KafkaUtils.prepareKafkaTopic(topic, processObjectDependencies)
+  override def initialParameters: List[Parameter] = Nil
 
-    val valueSchemaDeterminer = new SpecificRecordEmbeddedSchemaDeterminer(classTag[T].runtimeClass.asInstanceOf[Class[_ <: SpecificRecord]])
-    val valueSchemaData = valueSchemaDeterminer.determineSchemaUsedInTyping.valueOr(SchemaDeterminerErrorHandler.handleSchemaRegistryErrorAndThrowException)
-    val valueSchemaUsedInRuntime = valueSchemaDeterminer.toRuntimeSchema(valueSchemaData)
+  override def contextTransformation(context: ValidationContext, dependencies: List[NodeDependencyValue])(implicit nodeId: NodeId): NodeTransformationDefinition =
+    topicParamStep orElse {
+      case step@TransformationStep((TopicParamName, DefinedEagerParameter(topic:String, _)) :: Nil, _) =>
+        val preparedTopic = prepareTopic(topic)
+        val versionOption = LatestSchemaVersion
```

Comment:
Can you replace overloading of `prepareValueSchemaDeterminer(topics, version)` with fixed `LatestSchemaVersion` with inline creation of determiner as it was before? `LatestSchemaVersion` is bad because we want to use schema that is inside generated classes and it can be other than latest.

Reply:
also after changes with passing of schemadata in state I see that `SpecificRecordEmbeddedSchemaDeterminer` is not needed anymore - you just pass schemadata

Reply:
Brought back the inline creation. Added some comment with explanation.